jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java

```diff
@@ -1644,30 +1647,100 @@ public static String addQueries(String query1, String query2)
+        if (colon < 0)
+            return uri; // path portion not found
+
+        int end = -1;
+        if (rawURI.charAt(colon + 2) != '/')
+            end = colon + 2;
+        if (end >= 0)
+            return URI.create(rawURI.substring(0, colon) + ":///" + rawURI.substring(end));
+        return uri;
+    }
+
+    /**
+     * Take a URI and add a deep reference {@code jar:file://foo.jar!/suffix}, replacing
+     * any existing deep reference on the input URI.
+     *
+     * @param uri the input URI (supporting {@code jar} or {@code file} based schemes
+     * @param encodedSuffix the suffix to set.  Must start with {@code !/}.  Must be properly URI encoded.
+     * @return the {@code jar:file:} based URI with a deep reference
+     */
+    public static URI uriJarPrefix(URI uri, String encodedSuffix)
+    {
+        if (uri == null)
+            throw new IllegalArgumentException("URI must not be null");
+        if (encodedSuffix == null)
+            throw new IllegalArgumentException("Encoded Suffix must not be null");
+        if (!encodedSuffix.startsWith("!/"))
+            throw new IllegalArgumentException("Suffix must start with \"!/\"");
+
+        String uriString = uri.toASCIIString(); // ensure proper encoding
+
+        int bangSlash = uriString.indexOf("!/");
+        if (bangSlash >= 0)
+           uriString = uriString.substring(0, bangSlash);
+
+        if (uri.getScheme().equalsIgnoreCase("jar"))
+        {
+            return URI.create(uriString + encodedSuffix);
+        }
+        else if (uri.getScheme().equalsIgnoreCase("file"))
         {
-            if (!"jar".equals(uri.getScheme()))
-                return uri;
-            // Get SSP (retaining encoded form)
-            String s = uri.getRawSchemeSpecificPart();
-            int bangSlash = s.indexOf("!/");
-            if (bangSlash >= 0)
-                s = s.substring(0, bangSlash);
-            return new URI(s);
+            return URI.create("jar:" + uriString + encodedSuffix);
         }
-        catch (URISyntaxException e)
+        else
         {
-            throw new IllegalArgumentException(e);
+            throw new IllegalArgumentException("Unsupported URI scheme: " + uri);
         }
     }
 
-    public static String getJarSource(String uri)
+    /**
+     * Stream the {@link URLClassLoader#getURLs()} as URIs
+     *
+     * @param urlClassLoader the classloader to load from
+     * @return the Stream of {@link URI}
+     */
+    public static Stream<URI> streamOf(URLClassLoader urlClassLoader)
     {
-        if (!uri.startsWith("jar:"))
-            return uri;
-        int bangSlash = uri.indexOf("!/");
-        return (bangSlash >= 0) ? uri.substring(4, bangSlash) : uri.substring(4);
-    }
+        URL[] urls = urlClassLoader.getURLs();
+        return Stream.of(urls)
+            .filter(Objects::nonNull)
+            .map(URIUtil::getJarSource)
+            .map(URIUtil::fixBadJavaIoFileUrl);
```

Comment:
I'm not convinced calling `fixBadJavaIoFileUrl` here has any value.

Reply:
This fixes non-Jetty URLClassLoaders to produce to-spec URIs.
Perhaps it should be `ensureToSpecFileUri()`? 
But we can address the naming in the FileID/URIUtil second PR (not the first)